src/render/gif.rs

```diff
@@ -0,0 +1,277 @@
+        }
+
+        1
+    }
+
+    pub fn read_frame(frame: &Frame, gif_height: i32, gif_width: i32) -> Vec<u8> {
+        let median_color = Self::calculate_median_color_value(frame, gif_height, gif_width);
+
+        let mut image = Vec::new();
+        let mut buf: u8 = 0;
+
+        //the u64 is just in case someone put a gif that's huge (in terms of
+        // resolution), it shouldn't break
+        let width = u64::from(frame.width);
+        let height = u64::from(frame.height);
+
+        let pixels = &frame.buffer;
+
+        for y in 0..gif_height {
+            //if y is outside of the gif width
+            if y as u64 >= height {
+                continue;
+            }
+
+            //if y is outside of the screen
+            if y >= DISPLAY_HEIGHT {
+                continue;
+            }
+            for x in 0..gif_width {
+                //since we're using an array of u8, every 8 bit we need to start with a new int
+                if x % 8 == 0 && x != 0 {
+                    image.push(buf);
+                    buf = 0;
+                }
+                //if x is outside of the gif width
+                if x as u64 >= width {
+                    continue;
+                }
+
+                //if x is outside of the screen
+                if x >= DISPLAY_WIDTH {
+                    continue;
+                }
+
+                //calculating the index
+                let start: u64 = ((y as u64) * width + (x as u64)) * 4;
+
+                //getting the value of the pixels
+                let pixel_r = pixels.get(start as usize).unwrap_or(&0);
+                let pixel_g = pixels.get((start + 1) as usize).unwrap_or(&0);
+                let pixel_b = pixels.get((start + 2) as usize).unwrap_or(&0);
+
+                let mean = pixel_r / 3 + pixel_g / 3 + pixel_b / 3;
+                //I'm not sure if we should do something with the alpha channel of the gif
+                //I decided not to, but maybe we should
+
+                if mean >= median_color {
+                    //which bit to turn on
+                    let shift = x % 8;
+                    buf = buf + (128 >> shift);
+                }
+            }
+            //we fortcibly push the frame to the buffer after each line
+            image.push(buf);
+            buf = 0;
+        }
+        return image;
+    }
+
+    pub fn new(origin: Point, stop: Point, file: File) -> Self {
+        let gif_height = stop.y - origin.y;
+        let gif_width = stop.x - origin.x;
+
+        let mut decoder = gif::DecodeOptions::new();
+
+        decoder.set_color_output(gif::ColorOutput::RGBA);
+
+        let decoder_result = std::panic::catch_unwind(|| decoder.read_info(file).unwrap());
+
+        let mut decoded_frames = Vec::new();
+        let mut delays = Vec::new();
+        //this is to handle juste in case the file isn't a gif, or can't be parsed
+        // correctly
+        match decoder_result {
+            Ok(_) => {
```

Comment:
`if let Ok(_) = decoder_result` or even just `decoder_result.is_ok()` is probably more readable here